doc/neps/nep-0050-scalar-promotion.rst

```diff
@@ -0,0 +1,743 @@
+
+This NEP proposes to refactor the behaviour around two guiding principles:
+
+1. Values must never influence result type.
+2. NumPy scalars and 0-D arrays should behave consistently with their
+   N-D counterparts.
+
+We propose to remove all value-based logic and add special handling for
+Python scalars to preserve some convenient behaviors.
+Python scalars will be considered "weakly" typed.
+When a NumPy array/scalar is combined with a Python scalar, it will
+be converted to the NumPy dtype, such that::
+
+    np.array([1, 2, 3], dtype=np.uint8) + 1  # returns a uint8 array
+    np.array([1, 2, 3], dtype=np.float32) + 2.  # returns a float32 array
+
+There will be no dependence on the Python value itself.
+
+The proposed changes also apply to ``np.can_cast(100, np.int8)``, however,
+we expect that the behaviour in functions (promotion) will, in practice, be far
+more important than the casting change itself.
+
+
+Schema of the new proposed promotion rules
+------------------------------------------
+
+After the change, the promotions in NumPy will follow the schema below.
+Promotion always occurs along the green lines:
+from left to right within their kind and to a higher kind only when
+necessary.
+The result kind is always the largest kind of the inputs.
+Note that NumPy considers ``float32`` to have a lower precision than ``int32``
+and ``uint32``, but makes an exception for ``float64`` (and higher).
+
+The Python scalars are inserted at the very left of each "kind" and the
+Python integer does not distinguish signed and unsigned.
+When the promoting a Python scalar with a dtype of lower kind
+category (boolean, integral, inexact) with a higher one, we  use the
+minimum/default precision: that is ``float64``, ``complex128`` or ``int64``
+(``int32`` is used on some systems, e.g. windows).
+
+.. figure:: _static/nep-0050-promotion-no-fonts.svg
+    :figclass: align-center
+
+See the next section for examples which clarify the proposed behavior.
+Further examples with a comparison to the current behavior can be found
+in the table below.
+
+Examples of new behaviour
+-------------------------
+
+Since the schema and logic are difficult to read with respect to some cases,
+these are examples of the new behaviour.  Below, the Python integer has no
+influence on the result type::
+
+    np.uint8(1) + 1 == np.uint8(2)
+    np.int16(2) + 2 == np.int16(4)
+
+In the following the Python ``float`` and ``complex`` are "inexact", but the
+NumPy value is integral, so we use at least ``float64``/``complex128``::
+
+    np.uint16(3) + 3.0 == np.float64(6.0)
+    np.int16(4) + 4j == np.complex128(4+4j)
+
+But this does not happen for ``float`` to ``complex`` promotions, where
+``float32`` and ``complex64`` have the same precision::
+
+    np.float32(5) + 5j == np.complex64(5+5j)
+
+Note that the schema omits, ``bool``.  It is set below "integral", so that the
+following hold::
+
+    np.bool_(True) + 1 == np.int64(2)
```

Comment:
Hmmm, certainly also plausible.  My gut feeling is that this is the right choice, because:
```
np.int8(3) + 1.
```
should be float64 I think, since there is no clear logic to prefer float32 (or float16?).